plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,648 @@
+{
+	FuCapCmd cmd;
+	const guint32 dataszbyte = data_len * FU_SYNAPTICS_CAPE_WORD_IN_BYTES;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	memset(&cmd, 0, sizeof(cmd));
+	cmd.cmd_id = cmd_id;
+	cmd.module_id = module_id;
+
+	if (data_len != 0 && data != NULL) {
+		cmd.data_len = data_len;
+		if (!fu_memcpy_safe((guint8 *)cmd.data,
+				    sizeof(cmd.data),
+				    0, /* dst */
+				    (const guint8 *)data,
+				    dataszbyte,
+				    0, /* src */
+				    dataszbyte,
+				    error))
+			return FALSE;
+	}
+	return fu_synaptics_cape_device_sendcmd_ex(self, &cmd, 0, error);
+}
+
+static gboolean
+fu_synaptics_cape_device_sendcmd_ex(FuSynapticsCapeDevice *self,
+				    FuCapCmd *req,
+				    gulong delay_us,
+				    GError **error)
+{
+	FuCapCmdHidReport report;
+	guint elapsed_ms = 0;
+
+	/*First two bytes are report id.*/
+	report.report_id = GINT16_TO_LE(FU_SYNAPTICS_CAPE_DEVICE_GOLEM_REPORT_ID);
+
+	memcpy(&report.cmd, req, sizeof(report.cmd));
+
+	/*Set data length to MAX for any GET commands*/
+	if (FU_SYNAPTICS_CMD_GET_FLAG & report.cmd.cmd_id) {
+		report.cmd.data_len = GUINT32_TO_LE(FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN);
+	}
+	report.cmd.data_len = GUINT32_TO_LE(report.cmd.data_len);
+	report.cmd.cmd_id = GUINT32_TO_LE(report.cmd.cmd_id);
+	report.cmd.module_id = GUINT32_TO_LE(report.cmd.module_id);
+
+	if (!fu_synaptics_cape_device_set_report(self,
+						 (const guint8 *)&report,
+						 sizeof(report),
+						 error)) {
+		g_prefix_error(error, "failed to send: ");
+		return FALSE;
+	}
+	if (delay_us > 0)
+		g_usleep(delay_us);
+
+	/* wait for the command to complete */
+	for (; elapsed_ms < FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_TIMEOUT;
+	     elapsed_ms += FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL) {
+		if (!fu_synaptics_cape_device_get_report(self,
+							 (guint8 *)&report,
+							 sizeof(report),
+							 error))
+			return FALSE;
+		if (report.cmd.reply)
+			break;
+		g_usleep(FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL * 1000);
+	}
+
+	report.cmd.data_len = GUINT32_FROM_LE(report.cmd.data_len);
+
+	/* copy returned data if it is GET command*/
+	if (FU_SYNAPTICS_CMD_GET_FLAG & report.cmd.cmd_id && report.cmd.reply) {
+		memcpy(req, &report.cmd, sizeof(report.cmd));
+	}
+
+	return fu_synaptics_cape_device_rc_set_error(&report.cmd, error);
+}
+
+gboolean
+fu_synaptics_cape_device_setup_version(FuSynapticsCapeDevice *self, GError **error)
+{
+	guint32 version_raw;
+	FuCapCmd cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	cmd.cmd_id = GUINT32_TO_LE(FU_SYNAPTICS_CMD_GET_VERSION);
+	cmd.module_id = FU_SYNAPTICS_CAPE_CMD_APP_ID('C', 'T', 'R', 'L');
+	cmd.data_len = 4;
+
+	fu_synaptics_cape_device_sendcmd_ex(self, &cmd, 0, error);
+	if (self->version) {
+		g_free(self->version);
+	}
+	version_raw = (cmd.data[0] & 0xff) << 24 | (cmd.data[1] & 0xff) << 16 |
+		      (cmd.data[2] & 0x8) << 24 | (cmd.data[3] & 0xff);
+
+	self->version = fu_common_version_from_uint32(version_raw, FWUPD_VERSION_FORMAT_QUAD);
```

Comment:
why save locally as well as in FuDevice?

Reply:
I am not aware of built-in 'version', will remove this local copy.

Reply:
You'll have to generate a local version, e.g. `g_autofree gchar *version = NULL` but then we can throw it away :)